### PR TITLE
intel_mp app instantiations make stats symlink counter forest

### LINKED ISSUE #1178: Fix intel_mp RRD counter collection
When we made the 2018.06.01 release, we had some issues with intel_mp counters that had just been resolved prior to the release.  Unfortunately that refactoring broke RRD recording for intel_mp counters.

The fundamental issue is that the ptree manager uses inotify (or periodic polling) to see what counters are present for its set of workers.  However the symlink mechanism meant that the intel counters are no longer under the workers, and instead in `/var/run/snabb/intel-mp`.  As that directory wasn't being inotified, we didn't see its counters, and thusly the ptree manager failed to create RRD files.

Now, we could monitor `/var/run/snabb/intel-mp`.  However that has a problem, in that it could be there are other instances there; we only want to monitor the general statistics and the queue-specific statistics of the stats queue for the ptree workers.

I propose to fix this by making the symlink tree a bit deeper -- instead of symlinking directories, we only symlink files (in practice, .counter files).  If we need a directory, we create it under the app instead of symlinking.  That way inotify will see things automatically.

This will need some refactors in "snabb top" as well.

### PR DESCRIPTION
**Pull request by wingo**
This commit changes the intel_mp driver to make a forest of symlinks to the shared stats counters, rather than a single link to the shared counters directory.  This allows the app to indicate which queue it's on, as well as allowing ptree and top to naturally track the set of counters used by the app so they can manage associated .rrd files.

As part of this change, we remove the instance-local "pci" directory, instead using per-app "pci" directories.  This allows us to more nicely record which app is on which device.

This commit also updates "snabb top" to get PCI interfaces from within apps' SHM frames.

Fixes #1178.

WDYT @takikawa @dpino ?

### CODE REVIEW

**Comment by wingo** [comment]
By the way, when you skim back in time for a network function that's live and receiving traffic, it shows some slight counter fluctuation.  This is because the last-sample time keeps advancing, I think.  It's a bit confusing and should be fixed eventually but it's not harmful.

File: src/core/shm.lua [hunk 1]
```diff
@@ -63,8 +63,7 @@ end
 
 function alias (name, target)
    mkdir(lib.dirname(resolve(name)))
-   assert(S.symlink(root.."/"..resolve(target), root.."/"..resolve(name)),
-          "shm alias failed")
+   assert(S.symlink(root.."/"..resolve(target), root.."/"..resolve(name)))
```
Comment (eugeneia) [inline src/core/shm.lua:66]: Any reason for removing the assertion error message?

**Comment by eugeneia** [comment]
IIUC this will require the old `snabb top` (new `snabb ps`) to add special casing to read device statistics?  Currently, it is possible to list device statistics via `snabb top -l pci/<addr> <pid>`. Have to think about how to query non-process-bound resources...

File: src/apps/intel_mp/intel_mp.lua [hunk 2]
```diff
@@ -452,6 +509,41 @@ function Intel:new (conf)
    return self
 end
 
+function Intel:create_stats_symlinks ()
```
Comment (dpino) [inline src/apps/intel_mp/intel_mp.lua]: It seems this function is not used, when are the symlinks created?

**Comment by eugeneia** [comment]
On another note, I think having an intel_mp specific directory is a step back on our quest for driver consolidation.

How about the directory was called `/var/run/snabb/pci`, and master processes would symlink `/var/run/snabb/pci/<addr> -> /var/run/snabb/<pid>/group/pci/<addr>`? I think that would solve both of my concerns.

Edit: Actually, if the directory name is required for the new `snabb top` to deduce the pci/app relationship, I would even be fine if its `/var/run/snabb/intel-mp/<addr>` as long as its linked to `.../<pid>/group/pci/<addr>`.

**Comment by takikawa** [comment]
It looks like `get_rxstats` / `get_txstats` won't quite work with these changes since the shm frame is moved from a `pci/...` folder to an app-specific one. The other thing is that the app-specific shm frame doesn't have working entries for shared counters since the `create` method doesn't return an appropriate object for reading. PR #1180 might be a way to fix this.

File: src/apps/intel_mp/intel_mp.lua [hunk 3]
```diff
@@ -803,6 +895,7 @@ function Intel:stop ()
       self:unset_pool()
    end
    self:unset_tx_rate()
+   self:remove_stats_symlinks()
```
Comment (takikawa) [inline src/apps/intel_mp/intel_mp.lua]: I think some tests are failing due to this function call. In particular because `self.shm_frame` isn't defined (also above in `create_stats_symlinks`). Tests seem to make progress with this call removed though.

File: src/core/shm.lua (hunk 1 above)
Comment (wingo) [reply src/core/shm.lua:66]: S.symlink returns two values in an error case, but passing two args to `assert` truncates to the first value.  Removing the message allows the original error to propagate through.

**Comment by wingo** [comment]
@eugeneia regarding "snabb ps", I don't know that there is any significant change -- you get the statistics under the app/ directory of the Snabb instance, and also in the global intel-mp directory, but not in the per-Snabb-instance pci/ directory.

**Comment by wingo** [comment]
I have no idea why these tests are failing :(((  Deep flakiness :(((

**Comment by takikawa** [comment]
@wingo Re: the latest test failures, my theory after poking around a bit is that one problem is with the `return counter.open(target(name))` line in my commit 79d45461171f4ee34b4d0ed064e1f8941b7cd87c.

The issue with that call is that the target may not actually be created yet, if the snabb apps get stopped and restarted (as happens in the snabbnfv selftest).

For example, say we restart the stat collecting NIC app. When it closes, it will delete the shm frame for the stats and relinquish its stat collection status. When it comes back up, it may see that the lock is still held for the NIC (because of another NIC app) and won't start back up as stat collector.

In that case, the `self.shm` counters can't actually open the symlinked targets (since they were deleted and not recreated) and the calls fail.

Basically, the interaction between stat collection and restarting apps seems messy right now. I think we could change the line to only open the counter if it `shm.exists()` (and otherwise return an object that will error when used or something), or maybe make `get_rxstats` read the counter a different way.

**Comment by wingo** [comment]
Latest commit avoids eagerly opening the counters.  Let's see what happens!

**Comment by wingo** [comment]
yes!!! @takikawa wdyt?

**Review by takikawa — APPROVED** [review]
